stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,844 @@
+/// This software is distributed under the terms of the MIT License.
+/// Copyright (c) 2016-2020 UAVCAN Development Team.
+/// Authors: Pavel Kirienko <pavel.kirienko@zubax.com>, Tom De Rybel <tom.derybel@robocow.be>
+
+#include "bxcan.h"
+#include "bxcan_registers.h"
+#include <assert.h>
+#include <string.h>
+
+/// Configure the maximum interface index for the bxCAN hardware available in your MCU.
+/// Must be set to either 0 (only CAN1) or 1 (CAN1 and CAN2).
+#if !defined(BXCAN_MAX_IFACE_INDEX)
+#    error "Please set BXCAN_MAX_IFACE_INDEX to the maximum index of the available bxCAN hardware in your MCU."
+#endif
+
+/// Configure the system core clock frequency in Hz.
+/// Only used by the busy wait in waitMSRINAKBitStateChange().
+#if !defined(BXCAN_BUSYWAIT_DELAY_SYSTEM_CORE_CLOCK)
+#    error "Please set BXCAN_BUSYWAIT_DELAY_SYSTEM_CORE_CLOCK to the current system core clock."
+#endif
+
+/// By default, this macro resolves to the standard assert(). The user can redefine this if necessary.
+/// To disable assertion checks completely, make it expand into `(void)(0)`.
+#ifndef BXCAN_ASSERT
+// Intentional violation of MISRA: assertion macro cannot be replaced with a function definition.
+#    define BXCAN_ASSERT(x) assert(x)  // NOSONAR
+#endif
+
+/// Trigger an assertion failure if an inner priority inversion is detected at runtime.
+/// This setting has no effect if NDEBUG is defined (i.e., assertion checks disabled).
+#if !defined(BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION)
+#    define BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION true
+#endif
+
+/// Bit mask for the 29-bit extended ID data frame format.
+#define BXCAN_FRAME_EXT_ID_MASK 0x1FFFFFFFU
+
+/// Error trackers, one for each possible CAN interface.
+/// As the bxCANReapError() function only returns a bool, we can keep this as a simple flag.
+static bool g_error[2] = {false};
+
+/// Time stamps for TX time-out management. There are three TX mailboxes for each CAN interface.
+/// An UINT64_MAX value means the time stamp is not set (mailbox is not in use).
+static uint64_t g_tx_deadline[(1 + BXCAN_MAX_IFACE_INDEX) * 3] = {UINT64_MAX};
```

Comment:
Confusingly, the syntax you used prescribes that the first element is to be initialized as `UINT64_MAX`, and all other elements are to be zeroed. Details: https://en.cppreference.com/w/c/language/array_initialization

If you want non-zero values, you have to specify a dedicated value for each element in the array, which would be awkward considering that the length of the array is determined by the preprocessor here. Maybe it's better to drop initialization here (statics are zero-initialized by default) and assign the initial values in `bxCANConfigure()`?

Reply:
Got it. As memset is only for bytes, I've added a for-loop to set the array. (Actually I removed it again, see below).

The way the code is structured, it doesn't matter much to which value they are initialized, provided that all mailboxes are free when bxCAN is first enabled. Zero initialization may actually be more robust, as this will trigger the test in abortExpiredTxMailboxes() when sending data in bxCANPUSH(), where a busy mailbox (due to some strangeness at start-up) will get aborted. If it starts in the idle state (UINT64_MAX or some other very high value), yet is busy, such abort will not happen.

bxCANPush() does not care what the time-out value is, and will over-write it with a new one, as long as the mail box is free. (It only checks the hardware flags for availability.)

Thus, thinking about it again, I've changed it to zero initialization, as this does seem more robust.